File: AllTasks/DataStructures/Stack.pas
uses crt;

type
  PNode = ^Node;
  Node = record
    data: integer;
    next: PNode;
  end;

var
  expr: string;
  S: PNode;
  number, err: integer;

procedure Push(var Head: PNode; x: integer);
var
  NewNode: PNode;
begin
  New(NewNode); //Выделяем память - указываем в памяти на адрес структуры
  NewNode^.data := x; //Записываем значение в указанной памяти
  NewNode^.next := Head; //Сделать первым узлом
  Head := NewNode;
end;

procedure Pop(var Head: PNode);
var
  q: PNode;
begin
  while true do //Цикл нужен для обработки ошибки, если введен '-', когда еще не был создан узел
  begin
    if (Head = nil) then break;
    q := Head; //запомнить вершину
    Head := Head^.next; //удалить вершину из стека
    Dispose(q); //удалить из памяти
    break;
  end;
end;

function isEmpty(S: PNode): Boolean;
begin
  Result := (S = nil);
end;

procedure show(S: PNode);//Вывод данных из стека, начиная с нижнего элемента
var
  i: integer; arr: array of integer;
begin
  i := 0;;
  while S <> nil do
  begin
    setlength(arr, i + 1);
    arr[i] := s^.data;
    s := s^.next;
    i += 1;
  end;
  for i := length(arr) - 1 downto 0 do print(arr[i]);
end;

begin
  S := nil;
  while not EOF(F) do
  begin
    readln(f, expr);
    if (expr[1] = '+') then
    begin
      delete(expr, 1, 1);
      val(expr, number, err);
      if (expr = '') then begin clrscr; continue; end; //Исключает случай, если введен просто плюс
      push(S, number)
    end
    else if (expr[1] = '-') then
    begin
      pop(S);
      if isEmpty(S) then
      begin
        writeln('Стек пуст');
        break;
      end
    end else break;
  end;
  show(S);
end.

File: AllTasks/DataStructures/Stack.cpp
#include <iostream>

using namespace std;

struct List
{
	int x; //информационный элемент
	List *Next, *Head; //Голова стека и указатель на следующий элемент
};

/*ФУНКЦИЯ ДОБАВЛЕНИЯ ЭЛЕМЕНТА В СТЕК (в список LIFO)*/
void Add(int x, List *&MyList) //Принимаем элемент стека и указатель на стек, при этом говорим, что принимаемый указатель будет сам по себе указателем
{
	List *temp = new List; //Выделяем память для нового элемента
	temp->x = x; //Записываем в поле x принимаемый в функцию элемент x
	temp->Next = MyList->Head; //Указываем, что следующий элемент это предыдущий
	MyList->Head = temp; //Сдвигаем голову на позицию вперед
}

/*ФУНКЦИЯ ОТОБРАЖЕНИЯ СТЕКА*/
void Show(List *MyList) //Нужен только сам стек
{
	List *temp = MyList->Head; //Объявляем указатель и Указываем ему, что его позиция в голове стека
							   //с помощью цикла проходим по всему стеку
	while (temp != NULL) //выходим при встрече с пустым полем
	{
		cout << temp->x << " "; //Выводим на экран элемент стека
		temp = temp->Next; //Переходим к следующему элементу
	}
}

/*ФУНКЦИЯ УДАЛЕНИЯ СТЕКА ИЗ ПАМЯТИ*/
void ClearList(List *MyList)
{
	while (MyList->Head != NULL) //Пока по адресу не пусто
	{
		List *temp = MyList->Head->Next; //Временная переменная для хранения адреса следующего элемента
		delete MyList->Head; //Освобождаем адрес обозначающий начало
		MyList->Head = temp; //Меняем адрес на следующий
	}
}

int main()
{
	List *MyList = new List; //Выделяем память для стека
	MyList->Head = NULL; //Во избежание ошибок инициализируем первый элемент
	while (true)
	{
		int num;
		cin >> num;
		if (num == 1) break;
		Add(num, MyList); //Заносим данные в стек
	}
	Show(MyList); //Выводим стек на экран
	ClearList(MyList); //Очищаем память.
	delete MyList->Head;
	delete MyList;
	system("pause");
}
